src/index.rs

```diff
@@ -1,65 +1,105 @@
 use anyhow::anyhow;
-use canonical_path::{CanonicalPath, CanonicalPathBuf};
-use itertools::Itertools;
+use log;
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
 use std::collections::{HashMap, HashSet};
 use std::fs::{self, File, Metadata};
-use std::io::{BufRead, BufReader, Write};
-use std::ops::Add;
+use std::io::BufReader;
+use std::io::BufWriter;
 use std::path::{Path, PathBuf};
-use std::str::FromStr;
-use std::time::{Duration, SystemTime, UNIX_EPOCH};
+use std::time::{Duration, SystemTime};
 use walkdir::{DirEntry, WalkDir};
 
-use log;
+use crate::{id::ResourceId, ArklibError, Result, ARK_FOLDER, INDEX_PATH};
 
-use crate::id::ResourceId;
-use crate::{ArklibError, Result, ARK_FOLDER, INDEX_PATH};
+pub const RESOURCE_UPDATED_THRESHOLD: Duration = Duration::from_millis(1);
+pub type Paths = HashSet<PathBuf>;
 
-#[derive(Eq, Ord, PartialEq, PartialOrd, Hash, Clone, Debug)]
+/// IndexEntry represents a [`ResourceId`] and the time it was last modified
+#[derive(
+    Eq, Ord, PartialEq, PartialOrd, Hash, Clone, Debug, Serialize, Deserialize,
+)]
 pub struct IndexEntry {
+    /// The time the resource was last modified
     pub modified: SystemTime,
+    /// The resource's ID
     pub id: ResourceId,
 }
 
-#[derive(PartialEq, Clone, Debug)]
+/// Represents an index of resources stored as files
+/// in the filesystem under some prefix, or "root".
+///
+/// This struct maintains a mapping from resource IDs to their corresponding
+/// file paths, as well as the mapping of the opposite direction from file
+/// paths to index entries, which are simply resource IDs with modification
+/// timestamps.
+///
+/// The paths stored in the mappings are not absolute paths, but are relative
+/// to the "root". See methods `TODO` and `TODO` in the API, which can provide
+/// the end users with absolute/canonical(TODO) paths for the convenince.
+/// Otherwise, the end user is responsible for managing the working dir
```

Comment:
@tareknaser do you think we need functions to get absolute paths?

Actually, we could forbid querying particular paths because, ideally, the lib must be used in FRP way, i.e. only react to updates, which could be presented as a stream in the library wrappers. But it could be a bit tough for the end-users... If we adopt this way, the API would become cleaner though. We would need initialization parameter which specifies what kind of paths we emit in the updates: relative or absolute. This would become the preference of the end-user.

If we provide only "reactive API", then I'm not sure what to do with `track` functions.

Reply:
I think the default should always be that we are dealing with relative path. The `root_path` is a field in `ResourceIndex` and any relative path can be prepended with it to get the canonical path. 

We just have to make sure that `root_path` is canonical on its own. We currently do this in `ResourceIndex::build()` but we also need a getter for it

Reply:
Agree, and also we must ensure that the paths are canonicalized before trimming the common prefix.